copy/multiple_test.go

```diff
@@ -1,36 +1,78 @@
+	require.NoError(t, err)
+	list, err := internalManifest.ListFromBlob(validManifest, internalManifest.GuessMIMEType(validManifest))
+	require.NoError(t, err)
 	// Test CopyAllImages
 	sourceInstances := []digest.Digest{
-		digest.Digest("sha256:5b0bcabd1ed22e9fb1310cf6c2dec7cdef19f0ad69efa1f392e94a4333501270"),
+		digest.Digest("sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"),
 		digest.Digest("sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"),
 		digest.Digest("sha256:cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc"),
 	}
 
-	instancesToCopy := prepareInstanceCopies(sourceInstances, &Options{})
+	instancesToCopy, err := prepareInstanceCopies(list, sourceInstances, &Options{})
+	require.NoError(t, err)
 	compare := []instanceCopy{}
-	for _, instance := range sourceInstances {
+
+	// Compare all fields other than compressionVariant.
+	for i, instance := range sourceInstances {
+		defaultCompressionVariant := OptionCompressionVariant{Algorithm: instancesToCopy[i].compressionVariant.Algorithm}
 		compare = append(compare, instanceCopy{op: instanceCopyCopy,
-			sourceDigest: instance})
+			sourceDigest: instance, compressionVariant: defaultCompressionVariant})
 	}
 	assert.Equal(t, instancesToCopy, compare)
 
+	// Compare compressionVariants
+	expectedCompression := []string{"zstd", "gzip", "gzip"}
+	actualCompression := []string{}
+	for _, instance := range instancesToCopy {
+		actualCompression = append(actualCompression, instance.compressionVariant.Algorithm.Name())
+	}
+	assert.Equal(t, expectedCompression, actualCompression)
+
 	// Test with CopySpecific Images
 	copyOnly := []digest.Digest{
 		digest.Digest("sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"),
 	}
-	instancesToCopy = prepareInstanceCopies(sourceInstances, &Options{
+	instancesToCopy, err = prepareInstanceCopies(list, sourceInstances, &Options{
 		Instances:          copyOnly,
 		ImageListSelection: CopySpecificImages})
+	require.NoError(t, err)
+	assert.Equal(t, "gzip", instancesToCopy[0].compressionVariant.Algorithm.Name())
+	defaultCompressionVariant := OptionCompressionVariant{Algorithm: instancesToCopy[0].compressionVariant.Algorithm}
 	assert.Equal(t, instancesToCopy, []instanceCopy{{
-		op:           instanceCopyCopy,
-		sourceDigest: copyOnly[0]}})
+		op:                 instanceCopyCopy,
+		sourceDigest:       copyOnly[0],
+		compressionVariant: defaultCompressionVariant}})
+
+	// Test with CopySpecific Image with EnsureCompressionExists
+	copyOnly = []digest.Digest{
+		digest.Digest("sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"),
+	}
+	instancesToCopy, err = prepareInstanceCopies(list, sourceInstances, &Options{
+		Instances:                      copyOnly,
+		EnsureCompressionVariantsExist: []OptionCompressionVariant{{Algorithm: &compression.Zstd}},
+		ImageListSelection:             CopySpecificImages})
```

Comment:
It seems to me that without `CopySpecificImages`, a full investigation of the results would show that this causes unwanted clones. So that might also be a good test case. (And possibly, a test case with `Ensure…Exist: Gzip`, with the same source image.)